pytket/pytket/mbqc_compile/mpattern.py

```diff
@@ -0,0 +1,434 @@
+                    finished_edge_pool[v] = 0
+            vlist = [x for _, x in sorted(zip(dlist, vlist))]
+            vlist.reverse()
+            c.add_barrier(range(len(g.vertices())), range(len(g.vertices())))
+        return c
+
+    def remove_redundant(self, g: Graph) -> None:
+        """
+        Removes simples edges from a zx diagram by merging the connected
+        vertices.
+        
+        :param g:       A zx diagram.
+        :param type:    Graph
+        """
+        simple_edges = set()
+        for edge in g.edge_set():
+            if g.edge_type(edge)== 1:
+                simple_edges.add(edge)
+        for edge in simple_edges:
+            v1,v2 = edge
+            g.remove_edge(edge)
+            is_boundary = True
+            removing_input = True
+            remove_vertex = v1
+            keep_vertex = v2
+            if v1 in g.inputs:
+                remove_vertex = v1
+                keep_vertex = v2
+            elif v2 in g.inputs:
+                remove_vertex = v2
+                keep_vertex = v1
+            elif v1 in g.outputs:
+                removing_input = False
+            elif v2 in g.outputs:
+                removing_input = False
+                remove_vertex = v2
+                keep_vertex = v1
+            else:
+                is_boundary = False
+            g.set_row(keep_vertex, g.row(remove_vertex))
+            g.set_qubit(keep_vertex,g.qubit(remove_vertex))
+            neighbors = g.neighbors(remove_vertex) - [keep_vertex]
+            for neighbor in neighbors:
+                this_type = g.edge_type((remove_vertex, neighbor))
+                if g.connected(keep_vertex,neighbor):
+                    if (g.edge_type((keep_vertex, neighbor)) == this_type):
+                        g.remove_edge((keep_vertex,neighbor))
+                    else:
+                        g.add_edge((keep_vertex,neighbor),this_type)
+                else:
+                    g.add_edge((keep_vertex,neighbor),this_type)
+            g.add_to_phase(keep_vertex,g.phase(remove_vertex))
+            g.remove_vertex(remove_vertex)
+            if is_boundary:
+                if removing_input:
+                    g.inputs.remove(remove_vertex)
+                    g.inputs.append(keep_vertex)
+                    for i in self.inputs.keys():
+                        if self.inputs[i] == remove_vertex:
+                            self.inputs[i] = keep_vertex
+                            break
+                else:
+                    g.outputs.remove(remove_vertex)
+                    g.outputs.append(keep_vertex)
+                    for o in self.outputs.keys():
+                        if self.outputs[o] == remove_vertex:
+                            self.outputs[o] = keep_vertex
+                            break
+                g.set_type(keep_vertex, 0)
+        self.identity_cleanup(g)
+        pass
+
+    def identity_cleanup(self, g: Graph) -> None:
```

Comment:
Similarly does this kind of cleanup just not exist in pyzx?

Reply:
Basically using remove_redundant() might open the door to some further simplifications which identity_cleanup() takes care of. I can't pass it back to pyzx because remove_redundant() changes the format of the graphs slightly.